src/server.c

```diff
@@ -2716,9 +2721,55 @@ void makeThreadKillable(void) {
     pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
 }
 
-void initServer(void) {
-    int j;
+/* Return non-zero if the database is empty  */
+int databaseEmpty(int id) {
+    return id < 0 || id >= server.dbnum || !server.db[id] || kvstoreSize(server.db[id]->keys) == 0;
+}
+
+// /* Initialize temporary db on replica for use during diskless replication. */
+// serverDb *initTempDb(int id) {
+//     int slot_count_bits = 0;
+//     int flags = KVSTORE_ALLOCATE_HASHTABLES_ON_DEMAND;
+//     if (server.cluster_enabled) {
+//         slot_count_bits = CLUSTER_SLOT_MASK_BITS;
+//         flags |= KVSTORE_FREE_EMPTY_HASHTABLES;
+//     }
+//     serverDb *tempDb = zcalloc(sizeof(serverDb) * server.dbnum);
+//
+//     tempDb->id = id;
+//     tempDb->keys = kvstoreCreate(&kvstoreKeysHashtableType, slot_count_bits, flags);
+//     tempDb->expires = kvstoreCreate(&kvstoreExpiresHashtableType, slot_count_bits, flags);
+//
+//     return tempDb;
+// }
+serverDb *createDatabase(int id) {
+    int slot_count_bits = 0;
+    int flags = KVSTORE_ALLOCATE_HASHTABLES_ON_DEMAND;
+    if (server.cluster_enabled) {
+        flags |= KVSTORE_FREE_EMPTY_HASHTABLES;
+        slot_count_bits = CLUSTER_SLOT_MASK_BITS;
+    }
+
+    serverDb *db = zmalloc(sizeof(serverDb));
+    db->keys = kvstoreCreate(&kvstoreKeysHashtableType, slot_count_bits, flags);
+    db->expires = kvstoreCreate(&kvstoreExpiresHashtableType, slot_count_bits, flags);
+    db->expires_cursor = 0;
+    db->blocking_keys = dictCreate(&keylistDictType);
+    db->blocking_keys_unblock_on_nokey = dictCreate(&objectKeyPointerValueDictType);
+    db->ready_keys = dictCreate(&objectKeyPointerValueDictType);
+    db->watched_keys = dictCreate(&keylistDictType);
+    db->id = id;
+    db->avg_ttl = 0;
+    return db;
+}
 
+void initDatabase(int id) {
+    if (server.db[id] == NULL) {
+        server.db[id] = createDatabase(id);
+    }
+}
```

Comment:
In `kvstore.c` there is lazy creation of hashtables. It is similar to lazy creation of databases. In kvstore, a function called createHashtableIfNeeded is used in many places. It lazy-creates it and returns it.

I think we can use a similar pattern here. The name with "if needed" is also more clear than "init".

```suggestion
serverDb *createDatabaseIfNeeded(int id) {
    if (server.db[id] == NULL) {
        server.db[id] = createDatabase(id);
    }
    return server.db[id];
}
```

It can be used for example like this for the SELECT command:

```c
    c->db = createDatabaseIfNeeded(id);
```